### ISSUE TITLE
Support for private docker repositories/registries

### ISSUE THREAD

**Issue by cmosh** [exoframejs/exoframe#117]
It would be nice if there was support for private registries/repos

**Comment by yamalight** [comment]
I assume you mean support for installing templates from private repos?

**Comment by cmosh** [comment]
Yes I think so, but to be clear let me explain my current use case because I'm not entirely sure. Which is this:

 I have a private docker image on quay.io and I am using it as my base image in a Dockerfile I'd like to deploy.  Something like this:

`FROM quay.io/myaccount/mybaseimage
 RUN foo bar
 CMD ['run', 'foo','-bar']
`

But since the image is in a private docker registry the build fails, both the server I'm deploying to and my localhost are signed in and can pull the image, pre-pulling the image on the server does not help because exoframe tries to fetch the manifest on every build/deploy

**Comment by yamalight** [comment]
Ah, well, that can be easily fixed. All you need to do is to login into quay with your docker daemon.
E.g. ssh into your server and run `docker login quay.io`, enter your credentials and try to deploy a service again - it should work.

**Comment by cmosh** [comment]
As I've already explained I'm logged in on both my local machine and on the server, it was the first thought I had, and to test I even pulled the image onto the server directly but it still gives the same error : unauthorized etc.

**Comment by yamalight** [comment]
The fact that you cannot pull it with docker daemon directly means that you are not logged into quay.io. I'd suggest looking into their docs or contacting their support about that.
Once you are able to pull images using docker daemon, exoframe should work as well

**Comment by cmosh** [comment]
I have been saying that I am logged in with the docker daemon, I can pull images with docker daemon on the server. I have even pulled and launched the images manually but I cannot deploy with exoframe. I have even tried both gitlab.com registry and codefresh registry. 


I repeat

**Comment by yamalight** [comment]
Uff, sorry, I'm a bit under the weather today and might not get the problem from the first try. 
I think I finally understand the problem new. I tried running it myself and got the same problem.

What needs to be done is - Exoframe needs a way to store and add [header parameters](https://docs.docker.com/engine/api/v1.36/#operation/ImageBuild) during build execution that'd pass docker registry login and password.

**Comment by cmosh** [comment]
It's okay. Happens to all of us, I probably also wasn't as clear as I should have been

**Comment by yamalight** [comment]
@maxime1992 haven't had time to work on this at all (it's quite far down the list for me). Would be open to a PR 🤔

**Comment by maxime1992** [comment]
Damn :sweat_smile: 

I've been refactoring my deploy to avoid having that error:
https://github.com/exoframejs/exoframe/issues/199#issuecomment-457874817

So I've setup a full pipeline on GItlab to build the image first (as gitlab free runners are quite good even for private repos!) but now I can't pull it haha.

> haven't had time to work on this at all

No worries :) 

> Would be open to a PR

Would love to if I had some time for that which I definitely don't =(

**Comment by maxime1992** [comment]
Hi there, has anyone found a solution to this issue? (or even just a work around)

I'm still stuck :sweat_smile:

**Comment by alexanderadam** [comment]
There's [a pull request ongoing](https://github.com/exoframejs/exoframe-server/pull/19). I don't know about the current state of it though (and when it could be merged).

Also there's [a migration to Traefik 2.0](https://github.com/exoframejs/exoframe/issues/263) in the work. So I guess the code base _could_ lead to merge conflicts anyway.
